samples/subsys/usb/hid-cdc/src/main.c

```diff
@@ -0,0 +1,868 @@
+		u32_t bytes_read;
+
+		while ((bytes_read = uart_fifo_read(dev,
+		       (u8_t *)data_buf_kbd+chr_ptr_kbd,
+		       sizeof(data_buf_kbd)-chr_ptr_kbd))) {
+			chr_ptr_kbd += bytes_read;
+			if (data_buf_kbd[chr_ptr_kbd-1] == 0x0D) {
+				/* ENTER */
+				struct app_evt_t *ev = app_evt_alloc();
+
+				data_buf_kbd[chr_ptr_kbd-1] = '\0';
+				strcpy(string, data_buf_kbd);
+				ev->event_type = CDC_STRING;
+				flush_buffer_kbd();
+				app_evt_put(ev);
+				k_sem_give(&evt_sem);
+			}
+		}
+	}
+}
+
+/* Devices */
+
+static void left_button(struct device *gpio, struct gpio_callback *cb,
+			u32_t pins)
+{
+	struct app_evt_t *ev = app_evt_alloc();
+
+	ev->event_type = GPIO_BUTTON_0,
+	app_evt_put(ev);
+	k_sem_give(&evt_sem);
+}
+
+#ifdef SW1_GPIO_PIN
+static void right_button(struct device *gpio, struct gpio_callback *cb,
+			 u32_t pins)
+{
+	struct app_evt_t *ev = app_evt_alloc();
+
+	ev->event_type = GPIO_BUTTON_1,
+	app_evt_put(ev);
+	k_sem_give(&evt_sem);
+}
+#endif
+
+#ifdef SW2_GPIO_PIN
+static void x_move(struct device *gpio, struct gpio_callback *cb, u32_t pins)
+{
+	struct app_evt_t *ev = app_evt_alloc();
+
+	ev->event_type = GPIO_BUTTON_2,
+	app_evt_put(ev);
+	k_sem_give(&evt_sem);
+}
+#endif
+
+#ifdef SW3_GPIO_PIN
+static void y_move(struct device *gpio, struct gpio_callback *cb, u32_t pins)
+{
+	struct app_evt_t *ev = app_evt_alloc();
+
+	ev->event_type = GPIO_BUTTON_3,
+	app_evt_put(ev);
+	k_sem_give(&evt_sem);
+}
+#endif
+
+int callbacks_configure(struct device *gpio, u32_t pin, int flags,
+			void (*handler)(struct device*, struct gpio_callback*,
+			u32_t), struct gpio_callback *callback)
+{
+	if (!gpio) {
+		LOG_ERR("Could not find PORT");
+		return -ENXIO;
+	}
+	gpio_pin_configure(gpio, pin,
+			   GPIO_DIR_IN | GPIO_INT | GPIO_INT_DEBOUNCE |
+			   GPIO_INT_EDGE | flags);
+	gpio_init_callback(callback, handler, BIT(pin));
+	gpio_add_callback(gpio, callback);
+	gpio_pin_enable_callback(gpio, pin);
+	return 0;
+}
+
+void main(void)
+{
+	struct device *led_dev, *hid0_dev, *hid1_dev, *cdc0_dev, *cdc1_dev;
+	u32_t dtr = 0U;
+
+	caps = false;
+	chr_ptr_mouse = 0;
+	chr_ptr_kbd = 0;
+
+	/* Configure devices */
+
+	led_dev = device_get_binding(LED_PORT);
+	if (led_dev == NULL) {
+		LOG_ERR("Cannot get LED");
+		return;
+	}
+
+	hid0_dev = device_get_binding(CONFIG_USB_HID_DEVICE_NAME_0);
+	if (hid0_dev == NULL) {
+		LOG_ERR("Cannot get USB HID 0 Device");
+		return;
+	}
+
+	hid1_dev = device_get_binding(CONFIG_USB_HID_DEVICE_NAME_1);
+	if (hid1_dev == NULL) {
+		LOG_ERR("Cannot get USB HID 1 Device");
+		return;
+	}
+
+	cdc0_dev = device_get_binding(CONFIG_CDC_ACM_PORT_NAME_0);
+	if (cdc0_dev == NULL) {
+		LOG_ERR("Cannot get USB CDC 0 Device");
+		return;
+	}
+
+	cdc1_dev = device_get_binding(CONFIG_CDC_ACM_PORT_NAME_1);
+	if (cdc1_dev == NULL) {
+		LOG_ERR("Cannot get USB CDC 1 Device");
+		return;
+	}
+
+	gpio_pin_configure(led_dev, LED, GPIO_DIR_OUT);
+
+	if (callbacks_configure(device_get_binding(PORT0), PIN0, PIN0_FLAGS,
```

Comment:
All this code to configure the callbacks is copy-pasta, could you perhaps use an array?

Reply:
Each callback configuration (except for button 0) is called only if the button exists. This is to enable running this example on boards with only 1 button. See also `hid-mouse` example.